docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -5,99 +5,102 @@
-  H1 types.SignedHeader
-  H2 types.SignedHeader
+type ConflictingHeadersTrace struct {
+  Headers []*types.SignedHeader
 }
 ```
 
-_Remark_: Theoretically, only the header, which differs from what a full node
-has, needs to be sent. But sending two headers a) makes evidence easily
-verifiable b) simplifies the light client, which does not have query each
-witness as to which header it possesses.
-
-When a full node receives the `ConflictingHeadersEvidence` evidence, it should
+When a full node receives a `ConflictingHeadersTrace`, it should
 a) validate it b) figure out if malicious behaviour is obvious (immediately
-slashable) or the fork accountability protocol needs to be started.
+slashable) or run the amnesia protocol needs to be started.
 
 ### Validating headers
 
-Check both headers are valid (`ValidateBasic`), have the same height, and
-signed by 1/3+ of the validator set that the full node had at height
-`H1.Height`.
+Check headers are valid (`ValidateBasic`), are in order of ascending height, and do not exceed the `MaxTraceSize`.
+ 
+*NOTE: Do we need to limit the size of the header array if rpc already has a built in limiter on the size of the message you can send?* 
 
-- Q: What if light client validator set is not equal to full node's validator
-  set (i.e. from full node's point of view both headers are not properly signed;
-  this includes the case where none of the two headers were committed on the
-  main chain)
+### Finding Block Bifurcation
 
-  Reject the evidence. It means light client is following a fork, but, hey, at
-  least it will halt.
+The node pulls the block ID's for the respective heights of the trace headers and follows the lowest cost till rejection path.
 
-- Q: Don't we want to punish validators who signed something else even if they
-  have less or equal than 1/3?
+*Q: Is there a formal term for lowest-cost-till-rejection-path?*
 
-  No consensus so far. Ethan said no, Zarko said yes.
-  https://github.com/tendermint/spec/pull/71#discussion_r374210533
+First it checks to see that the first header hash matches its first `BlockID` else it can discard it. 
 
-### Figuring out if malicious behaviour is immediately slashable
+If the last header hash matches the nodes blockID then it can also discard it on the assumption that a fork can not remerge and hence this is just a trace of valid headers. 
 
-Let's say H1 was committed from this full node's perspective (see Appendix A).
-_If neither of the headers (H1 and H2) were committed from the full node's
-perspective, the evidence must be rejected._
+The node then continues to loop in descending order checking that the headers hash doesn't match it's own blockID for that height. Once it reaches the height that the block ID matches the hash it then sends the common header, the trusted header and the diverged header (common header is needed for lunatic evidence) to determine if the divergence is a real offense to the tendermint protocol or if it is just fabricated.
```

Comment:
Shouldn't this be ascending? It would start at the oldest header and iterate up the trace?

Reply:
So I don't know if there's too much difference between either way. I think basically you want to find the split point as soon as possible.  I don't know if there is any heuristic we can leverage where it is more likely that the split is going to be in the first half of the array or the second.